lib/Target/ExportSystemC/Patterns/FuncEmissionPatterns.cpp

```diff
@@ -0,0 +1,252 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This implements the emission patterns for the func dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "FuncEmissionPatterns.h"
+#include "../EmissionPrinter.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+
+using namespace mlir::func;
+using namespace circt;
+using namespace circt::ExportSystemC;
+
+//===----------------------------------------------------------------------===//
+// Operation emission patterns.
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// Emit a func.func function. Users of the function arguments request an
+/// expression to be inlined and we simply return the name of the argument.This
+/// name has to be passed to this emission pattern via an array of strings
+/// attribute called 'argNames' because the emitter cannot do any name uniquing
+/// as it just emits the IR statement by statement. However, relying on an
+/// attribute for the argument names also has the advantage that the names
+/// can be preserved during a lowering pipeline and upstream passes have more
+/// control on how the arguments should be named (e.g. when they create a
+/// function and have some context to assign better names).
+struct FuncEmitter : OpEmissionPattern<FuncOp> {
+  using OpEmissionPattern::OpEmissionPattern;
+  MatchResult matchInlinable(Value value) override {
+    if (!value.isa<BlockArgument>())
+      return {};
+
+    if (auto funcOp = value.getParentRegion()->getParentOfType<FuncOp>()) {
+      if (auto argNames = funcOp->getAttrOfType<ArrayAttr>("argNames")) {
+        if (argNames.size() == funcOp.getNumArguments() &&
+            llvm::all_of(argNames,
+                         [](Attribute arg) { return arg.isa<StringAttr>(); }))
+          return Precedence::VAR;
+      }
+    }
+
+    return {};
+  }
+
+  void emitInlined(Value value, EmissionPrinter &p) override {
+    auto func = value.getParentRegion()->getParentOfType<FuncOp>();
+    for (size_t i = 0, e = func.getNumArguments(); i < e; ++i) {
+      if (func.getArgument(i) == value) {
+        p << func->getAttr("argNames")
+                 .cast<ArrayAttr>()[i]
+                 .cast<StringAttr>()
+                 .getValue();
+        return;
+      }
+    }
+  }
+
+  bool matchStatement(Operation *op) override {
+    if (auto funcOp = dyn_cast<FuncOp>(op)) {
+      if (funcOp.getFunctionType().getNumResults() > 1)
+        return false;
+
+      if (funcOp.getNumArguments() == 0)
+        return true;
+
+      if (auto argNames = funcOp->getAttrOfType<ArrayAttr>("argNames"))
+        return argNames.size() == funcOp.getNumArguments() &&
+               llvm::all_of(argNames, [](Attribute arg) {
+                 return arg.isa<StringAttr>();
+               });
+    }
```

Comment:
This is kind of spooky. This is only able to emit func ops which meet a certain, higher standard. This seems slightly concerning as its overlaying a new level of verification on top of func ops which has a strange failure mode for users: if they don't properly populate a valid `argNames` array, then the emission just won't work.

I don't know how to handle this. 😕 This may hint that emitting `funcOp` is not the right way to go and we want something else that `funcOp` is converted to. That can be safely verified with things like `argNames` existing, being the right type, and having the right size.

Reply:
I agree that it is not very elegant (quite hacky actually) to just add attributes that are important for correctness of emission to an op that has no native support for it. Trying to add the features we need to the upstream `func.func` is no option since some things we need don't fit the upstream dialect. I'd suggest to create our own custom function operation that can encode all the features we need natively (as ODS attributes, etc.), namely
* argNames
* extern "C" support
* verifying that there is only one return value as in C/C++ functions
* etc.

I wouldn't call additional matching conditions "concerning as its overlaying a new level of verification" though. It's weird here because they check for some seemingly random attributes, but having several separate emission patterns for the same op is totally fine (e.g., one for functions with 0 return values, one for functions with 1 return value, one for functions with more than 1 return values) Although, I would recommend handling the latter case in a pre-pass and reduce it to one return value rather than dealing with it in the emitter. In that case it would just check that the pre-pass was executed and did all of its work and it's fine to fail if not. Aiming to cover all cases in the printer directly will just lead to one big monolithic blob of pass that's hard to understand, maintain, and reuse. But converting to a new operation just dedicated to represent (and verify) the subset of another operation supported by the printer (sitting at the end of the emission pipeline) may not be the best approach in all cases. In that sense, yes, the matching function of the pattern does some form of verification to check that the rest of the emission pipeline did it's work and nothing was forgotten there, but it's likely better than just introducing a new op to move that kind of verification to or to just crash. Sorry for the rambling 😅